fix broken avatar checks when server_name contains a port

ashfame opened the pull request:
This PR fixes `check_avatar_size_and_mime_type()` to successfully update avatars on homeservers running on non-default ports which it would mistakenly treat as remote homeserver while validating the avatar's size and mime type.

Essentially it was incorrectly using `host` as `server_name` inside of `check_avatar_size_and_mime_type()` without minding the port its actually running on.

Eg: If running synapse on `synapse.dev:8008` this function would infer the current homeserver url is `synapse.dev` and treat `synapse.dev:8008` as remote server while trying to fetch media's metadata to check for size and mime type. As a result, it fails to find the image itself & fails avatar validation & hence setting a new avatar.

Happy to address any feedback given so that we can land this change.

Signed-off-by: Ashish Kumar <ashfame@users.noreply.github.com>

clokep commented:
This looks reasonable, but seems like a thing that we would possibly break in the future again. Do you think a test-case could be added to `ProfileTestCase` for this?

ashfame commented:
@clokep Thanks for taking a look! I am not sure how I can write the test for `check_avatar_size_and_mime_type()` since that requires an actual valid mxc url since we pull out metadata by the mxc url and then we need two homeserver instances, one of which runs on non-default port.

In existing tests, I only saw dummy mxc urls like `mxc://test/good` and `mxc://test/small`

I am pretty new to both Python & Synapse, so I may be missing something obvious/established.

richvdh commented on synapse/handlers/profile.py:
```diff
@@ -307,7 +307,11 @@ async def check_avatar_size_and_mime_type(self, mxc: str) -> bool:
             return True
 
-        server_name, _, media_id = parse_and_validate_mxc_uri(mxc)
+        host, port, media_id = parse_and_validate_mxc_uri(mxc)
+        if port is not None:
+            server_name = host + ":" + str(port)
+        else:
+            server_name = host
```
As @clokep says, please can we have a test for this.

> I am not sure how I can write the test for check_avatar_size_and_mime_type() since that requires an actual valid mxc url since we pull out metadata by the mxc url and then we need two homeserver instances, one of which runs on non-default port.

Do we actually need two homeserver instances? it looks like we just call `get_cached_remote_media`, which just checks the database. Could you just make a call to `store_cached_remote_media` before trying to set the avatar?

ashfame replied:
Thanks for the pointer! I was able to create a unit test that passes with the change in this PR and fails without it.

richvdh commented on synapse/handlers/profile.py:
```diff
@@ -307,7 +307,11 @@ async def check_avatar_size_and_mime_type(self, mxc: str) -> bool:
-        server_name, _, media_id = parse_and_validate_mxc_uri(mxc)
+        host, port, media_id = parse_and_validate_mxc_uri(mxc)
+        if port is not None:
+            server_name = host + ":" + str(port)
```
As an aside: I'm really rather surprised that we allow setting the avatar to a remote server. What is your usecase for this?

ashfame replied:
Well, I am just running synapse locally on 8080 port and trying to set an avatar that I uploaded (so have a mxc url for it), but found this function was treating that as a remote server since it didn't mind the non-default port in use.

richvdh replied:
oh yes, of course

ashfame commented on changelog.d/13927.bugfix:
```diff
@@ -0,0 +1 @@
+Fixes avatar handling check on homeservers running on non-default ports. Contributed by @ashfame.
```
Done! `Fixes check_avatar_size_and_mime_type() to successfully update avatars on homeservers running on non-default ports which were mistakenly treated as remote homeserver while validating the avatar's size and mime type.`

richvdh commented on changelog.d/13927.bugfix:
```diff
@@ -0,0 +1 @@
+Fixes check_avatar_size_and_mime_type() to successfully update avatars on homeservers running on non-default ports which were mistakenly treated as remote homeserver while validating the avatar's size and mime type. Contributed by @ashfame.
```
Imperative mood, please; and users don't care which function you changed, just what the bug was.

```suggestion
Fix a bug which prevented setting an avatar on homeservers which have an explicit port in their `server_name`. Contributed by @ashfame.
```

(Does it affect *all* homeservers with ports in their server name, or is this only a problem with certain configuration options?)

ashfame replied:
> Does it affect all homeservers with ports in their server name

Yes!

richvdh replied:
Right. I was under the impression that `check_avatar_size_and_mime_type` only did anything if you had specified a `max_avatar_size` and/or a `allowed_avatar_mimetypes`. No?

ashfame replied:
Yes, my bad! You are right.

richvdh replied:
ok well, can you clarify this in the changelog entry please.

ashfame replied:
Done!

ashfame commented:
@richvdh Thanks for the feedback in this PR! I have addressed all the points and its ready for another review. ✌️ 

Regarding the failing checks, I am not sure what the problem is since the exact failing test pass when I run them locally:

```
--- PASS: TestMembersLocal (8.83s)
    --- PASS: TestMembersLocal/Parallel (0.00s)
        --- PASS: TestMembersLocal/Parallel/New_room_members_see_their_own_join_event (0.05s)
        --- PASS: TestMembersLocal/Parallel/Existing_members_see_new_members'_presence (0.05s)
        --- PASS: TestMembersLocal/Parallel/Existing_members_see_new_members'_join_events (0.06s)
PASS
```

richvdh commented on tests/handlers/test_profile.py:
```diff
@@ -327,6 +327,47 @@ def test_avatar_constraint_mime_type(self) -> None:
         self.assertFalse(res)
 
+    @unittest.override_config({"allowed_avatar_mimetypes": ["image/png"]})
+    def test_avatar_fetching_metadata_right_source(self) -> None:
+        """Tests that local and remote file are rightly fetched for metadata
+        when checking for avatar size and mime type"""
```
```suggestion
    def test_avatar_fetching_metadata(self) -> None:
        """Tests that local and remote files are correctly fetched for metadata
        when mime type restrictions are in place"""
```

richvdh commented on tests/handlers/test_profile.py:
```diff
@@ -327,6 +327,47 @@ def test_avatar_constraint_mime_type(self) -> None:
+                user_id=UserID.from_string("@whatever:test"),
+            )
+        )
+        self.get_success(
+            store.store_cached_remote_media(
+                media_id="remote",
+                media_type="image/png",
+                media_length=50,
+                origin=remote_server,
+                time_now_ms=self.clock.time_msec(),
+                upload_name=None,
+                filesystem_id="remote",
+            )
+        )
```
suggest putting this next to the code that is going to do the test (ie, *after* `self.handler.check_avatar_size_and_mime_type("mxc://test/local")`).

richvdh commented on tests/handlers/test_profile.py:
```diff
@@ -327,6 +327,47 @@ def test_avatar_constraint_mime_type(self) -> None:
+            )
+        )
+
+        res = self.get_success(
+            self.handler.check_avatar_size_and_mime_type("mxc://test/local")
+        )
+        self.assertTrue(res)
```
to test the fix properly, shouldn't the local server have a port in its name? You can set `"server_name"` in `@unittest.override_config`.

richvdh commented on tests/handlers/test_profile.py:
```diff
@@ -327,6 +327,47 @@ def test_avatar_constraint_mime_type(self) -> None:
+        )
+        self.assertTrue(res)
+
+        res = self.get_success(
+            self.handler.check_avatar_size_and_mime_type(
+                "mxc://" + remote_server + "/remote"
+            )
+        )
```
It would be good to add some comments about what exactly this is testing, and how. It looks like we are not attempting to simulate a fetch from a remote server? Is the idea that, if the media is cached, we shouldn't attempt to fetch it?

ashfame replied:
This should address all your remarks:

The issue I discovered was when non-default port is used for server, `check_avatar_size_and_mime_type` attempts to get metadata from wrong source - remote media cache instead of local.

Hence this test ensures function is looking at the right place. It doesn't care for whether `get_local_media()` or `get_cached_remote_media()` since that would be best handled by the tests written to test their output.

This is why I had the name of test as `test_avatar_fetching_metadata_right_source`. If this sounds good to you, I can add the same as comments in code.

richvdh replied:
yes, please add some comments to the code

ashfame replied:
Added!

ashfame commented:
@richvdh Is this ready to be merged now? Tests are failing because `Connection to files.pythonhosted.org timed out` in Complement tests.

richvdh commented on tests/handlers/test_profile.py:
```diff
@@ -329,8 +329,11 @@ def test_avatar_constraint_mime_type(self) -> None:
+        # This test only concerns itself with correctly figuring out the host and port
+        # based on mxc url so that metadata can be fetched from the right place i.e.
+        # get_local_media() for local image & get_cached_remote_media() for remote image
         remote_server = "test:8080"
```
put this definition next to where it is used?

richvdh commented on tests/handlers/test_profile.py:
```diff
@@ -344,6 +347,11 @@ def test_avatar_fetching_metadata_right_source(self) -> None:
                 user_id=UserID.from_string("@whatever:test"),
             )
         )
+        res = self.get_success(
+            self.handler.check_avatar_size_and_mime_type("mxc://test/local")
+        )
+        self.assertTrue(res)
```
again: this check seems to pass with no code changes. In order to check that we have fixed the original bug, you need to use a `server_name` with an explicit port.

richvdh commented on tests/handlers/test_profile.py:
```diff
@@ -329,8 +329,11 @@ def test_avatar_constraint_mime_type(self) -> None:
+        # get_local_media() for local image & get_cached_remote_media() for remote image
         remote_server = "test:8080"
         store = self.hs.get_datastores().main
 
```
more comments please.

```suggestion

        # First, check that we can validate metadata for *local* avatars,
        # even if the local server has an explicit port.
        # (This was previously a bug in Synapse.)
```

richvdh commented on tests/handlers/test_profile.py:
```diff
@@ -344,6 +347,11 @@ def test_avatar_fetching_metadata_right_source(self) -> None:
+        )
+        self.assertTrue(res)
+
         self.get_success(
```
```suggestion
        # Now also check it for media on remote servers with explicit ports.
        #
        # Note that `check_avatar_size_and_mime_type` never actively pulls
        # remote media, but instead relies on it already having been cached.
        # So, for the test, it is adequate to cache the remote media, and then
        # check that it can be found. 
        self.get_success(
```

richvdh commented on tests/handlers/test_profile.py:
```diff
@@ -329,8 +329,11 @@ def test_avatar_constraint_mime_type(self) -> None:
-        when checking for avatar size and mime type"""
+        """Tests that local and remote files are correctly fetched for metadata
+        when checking for avatar size and mime type."""
+        # This test only concerns itself with correctly figuring out the host and port
+        # based on mxc url so that metadata can be fetched from the right place i.e.
+        # get_local_media() for local image & get_cached_remote_media() for remote image
```
I'm afraid I don't really understand this comment.

I suggest you delete it and replace it with my suggestions below.

ashfame commented:
@richvdh Allow me to re-phrase the issue and the unit test:

`check_avatar_size_and_mime_type()` had a logical bug of incorrectly using `parse_and_validate_mxc_uri()` to figure out the `server_name`. Now the only test against that bug would be to check for the impact caused by the value of `server_name`, which is either `get_local_media()` gets called or `get_cached_remote_media()` gets called.

This is what the unit test tries to test, whether the expected method to retrieve metadata gets called or not, as a way to check whether we have the right server_name figured out or not. And since we pass it both: a local file first and then a remote file, we don't really need to test with an overriden config of server_name and the test works for both default and non-default ports of server_name in this manner.

Hope that explains my intention with this PR. I re-worked the test to make it more elaborative and much more descriptive.

Sorry for the back-and-forth on this issue and I am happy with any changes you want to make on top of it.

ashfame commented:
@richvdh Do you know when you will be able to take another look at this? I have simplified the test with lot more comments since your last review, as I put it in my 2 comments above.

richvdh commented:
@ashfame as it happens I was just looking at it. I've cleaned up the unit tests a bit more; if you are happy and the CI passes I will merge it.

richvdh commented:
Thanks for your contribution!
